torchgeo/trainers/detection.py

```diff
@@ -0,0 +1,255 @@
+                if parse(torchvision.__version__) >= parse("0.13"):
+                    if backbone_pretrained:
+                        kwargs["weights"] = BACKBONE_WEIGHT_MAP[
+                            self.hyperparams["backbone"]
+                        ]
+                    else:
+                        kwargs["weights"] = None
+                else:
+                    kwargs["pretrained"] = backbone_pretrained
+
+                backbone = resnet_fpn_backbone(**kwargs)
+            else:
+                raise ValueError(
+                    f"Backbone type '{self.hyperparams['backbone']}' is not valid."
+                )
+
+            anchor_generator = AnchorGenerator(
+                sizes=((32), (64), (128), (256), (512)), aspect_ratios=((0.5, 1.0, 2.0))
+            )
+
+            roi_pooler = MultiScaleRoIAlign(
+                featmap_names=["0", "1", "2", "3"], output_size=7, sampling_ratio=2
+            )
+            num_classes = self.hyperparams["num_classes"]
+            self.model = FasterRCNN(
+                backbone,
+                num_classes,
+                rpn_anchor_generator=anchor_generator,
+                box_roi_pool=roi_pooler,
+            )
+
+        else:
+            raise ValueError(
+                f"Model type '{self.hyperparams['detection_model']}' is not valid."
+            )
+
+    def __init__(self, **kwargs: Any) -> None:
+        """Initialize the LightningModule with a model and loss function.
+
+        Keyword Args:
+            detection_model: Name of the detection model type to use
+            backbone: Name of the model backbone to use
+            in_channels: Number of channels in input image
+            num_classes: Number of semantic classes to predict
+
+        Raises:
+            ValueError: if kwargs arguments are invalid
+        """
+        super().__init__()
+        # Creates `self.hparams` from kwargs
+        self.save_hyperparameters()  # type: ignore[operator]
+        self.hyperparams = cast(Dict[str, Any], self.hparams)
+
+        self.config_task()
+
+        self.val_metrics = MeanAveragePrecision()
+        self.test_metrics = MeanAveragePrecision()
+
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+        """Forward pass of the model.
+
+        Args:
+            x: tensor of data to run through the model
+
+        Returns:
+            output from the model
+        """
+        return self.model(*args, **kwargs)
+
+    def training_step(self, *args: Any, **kwargs: Any) -> Tensor:
+        """Compute and return the training loss.
+
+        Args:
+            batch: the output of your DataLoader
+
+        Returns:
+            training loss
+        """
+        batch = args[0]
+        x = batch["image"]
+        batch_size = x.shape[0]
+        y = [
+            {"boxes": batch["boxes"][i], "labels": batch["labels"][i]}
+            for i in range(batch_size)
+        ]
+        loss_dict = self.forward(x, y)
```

Comment:
According to https://github.com/microsoft/torchgeo/pull/790#discussion_r984943775 we should use `self()` instead of `self.forward()`. I'll submit a PR to change existing trainers, but can you make that change here?